File: eits_app/eits_app/report/eits_receipt_image/eits_receipt_image.py
# Copyright (c) 2025, GennextIT and contributors
# For license information, please see license.txt

# import frappe

# Copyright (c) 2024, Aditya and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    """
    Main function that ERPNext calls to generate the report
    Returns columns and data
    """
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    """
    Define the columns that will appear in the report
    Each column is a dictionary with field properties
    """
    return [
        {
            "fieldname": "name",
            "label": _("Receipt ID"),
            "fieldtype": "Link",
            "options": "Receipt EITS",
            "width": 140,
            "align": "left"
        },
        {
            "fieldname": "date",
            "label": _("Date"),
            "fieldtype": "Date",
            "width": 100,
            "align": "center"
        },
        {
            "fieldname": "bill_number",
            "label": _("Bill Number"),
            "fieldtype": "Int",
            "width": 120,
            "align": "left"
        },
        {
            "fieldname": "amountaed",
            "label": _("Amount (AED)"),
            "fieldtype": "Float",
            "width": 120,
            "align": "right"
        },
        {
            "fieldname": "paid_by",
            "label": _("Paid By"),
            "fieldtype": "Data",
            "width": 130,
            "align": "left"
        },
        {
            "fieldname": "paid_from",
            "label": _("Paid From"),
            "fieldtype": "Link",
            "options": "Customer",
            "width": 130,
            "align": "left"
        },
        {
            "fieldname": "custom_mode_of_payment",
            "label": _("Payment Method"),
            "fieldtype": "Link",
            "options": "Method of Payment",
            "width": 130,
            "align": "left"
        },
        {
            "fieldname": "image_count",
            "label": _("Images"),
            "fieldtype": "Int",
            "width": 70,
            "align": "center"
        },
        {
            "fieldname": "view_images",
            "label": _("View Images"),
            "fieldtype": "Data",
            "width": 130,
            "align": "center"
        },
        {
            "fieldname": "submit_action",
            "label": _("Submit"),
            "fieldtype": "Data",
            "width": 120,
            "align": "center"
        }
    ]

def get_data(filters):
    """
    Fetch the actual data for the report
    This query ensures we get unique receipt records with image counts
    """
    conditions = get_conditions(filters)
    
    # Main query to get unique receipt records with image count
    query = f"""
        SELECT 
            er.name,
            er.date,
            er.bill_number,
            er.amountaed,
            er.paid_by,
            er.paid_from,
            er.custom_mode_of_payment,
            er.docstatus,
            COUNT(CASE 
                WHEN ca.image IS NOT NULL 
                AND ca.image != '' 
                AND ca.image != '/files/' 
                THEN 1 
            END) as image_count,
            er.name as view_images,
            er.name as submit_action
        FROM `tabReceipt EITS` er
        LEFT JOIN `tabImage Attachments` ca 
            ON ca.parent = er.name 
            AND ca.parenttype = 'Receipt EITS'
            AND ca.docstatus != 2
        WHERE 
            er.docstatus != 2 
            {conditions}
        GROUP BY 
            er.name, 
            er.date, 
            er.bill_number, 
            er.amountaed, 
            er.paid_by, 
            er.paid_from, 
            er.custom_mode_of_payment,
            er.docstatus
        ORDER BY 
            er.date DESC, 
            er.creation DESC
    """
    
    # Execute query and return results
    data = frappe.db.sql(query, filters, as_dict=True)
    
    # Process the data to add any additional formatting if needed
    for row in data:
        # Ensure amount is properly formatted
        if row.get('amountaed'):
            row['amountaed'] = float(row['amountaed']) if row['amountaed'] else 0.0
        
        # Ensure image count is an integer
        row['image_count'] = int(row['image_count']) if row['image_count'] else 0
        
        # Format the view_images column - just show text that will be handled by JS
        if row['image_count'] > 0:
            row['view_images'] = "View Images"
        else:
            row['view_images'] = "No Images"
        
        # Format submit action based on docstatus
        if row.get('docstatus') == 0:  # Draft
            row['submit_action'] = "Submit"
        elif row.get('docstatus') == 1:  # Submitted
            row['submit_action'] = "Submitted"
        else:  # Cancelled
            row['submit_action'] = "Cancelled"
    
    return data

def get_conditions(filters):
    """
    Build WHERE clause conditions based on user filters
    """
    conditions = ""
    
    # Date range filter
    if filters.get("from_date"):
        conditions += " AND er.date >= %(from_date)s"
    
    if filters.get("to_date"):
        conditions += " AND er.date <= %(to_date)s"
    
    # Paid By filter (partial match)
    if filters.get("paid_by"):
        conditions += " AND er.paid_by LIKE %(paid_by)s"
        filters["paid_by"] = f"%{filters['paid_by']}%"
    
    # Paid From filter (exact match)
    if filters.get("paid_from"):
        conditions += " AND er.paid_from = %(paid_from)s"
    
    # Payment Method filter
    if filters.get("payment_method"):
        conditions += " AND er.custom_mode_of_payment = %(payment_method)s"
    
    # Bill Number filter (partial match)
    if filters.get("bill_number"):
        conditions += " AND er.bill_number LIKE %(bill_number)s"
        filters["bill_number"] = f"%{filters['bill_number']}%"
    
    # Only show records with images filter
    if filters.get("only_with_images"):
        conditions += """
            AND EXISTS (
                SELECT 1 FROM `tabImage Attachments` ca2 
                WHERE ca2.parent = er.name 
                AND ca2.parenttype = 'Receipt EITS'
                AND ca2.image IS NOT NULL 
                AND ca2.image != '' 
                AND ca2.image != '/files/'
                AND ca2.docstatus != 2
            )
        """
    
    return conditions

# ADD THIS NEW METHOD FOR SUBMITTING RECEIPT
@frappe.whitelist()
def submit_receipt(receipt_id):
    """
    Server method to submit a receipt record
    """
    try:
        # Check if receipt exists
        if not frappe.db.exists("Receipt EITS", receipt_id):
            frappe.throw(_("Receipt record not found"))
        
        # Check if user has write permission
        if not frappe.has_permission("Receipt EITS", "write", receipt_id):
            frappe.throw(_("You don't have permission to submit this receipt record"))
        
        # Get the document
        doc = frappe.get_doc("Receipt EITS", receipt_id)
        
        # Check if already submitted
        if doc.docstatus == 1:
            frappe.throw(_("Receipt is already submitted"))
        
        # Check if cancelled
        if doc.docstatus == 2:
            frappe.throw(_("Cannot submit a cancelled receipt"))
        
        # Submit the document
        doc.submit()
        
        frappe.db.commit()
        
        return {
            "success": True,
            "message": _("Receipt {0} has been submitted successfully").format(receipt_id),
            "docstatus": doc.docstatus
        }
        
    except Exception as e:
        frappe.db.rollback()
        frappe.log_error(frappe.get_traceback(), "Error in submit_receipt")
        frappe.throw(_("Error submitting receipt: {0}").format(str(e)))

# ADD THIS NEW METHOD FOR FETCHING IMAGES
@frappe.whitelist()
def get_receipt_images(receipt_id):
    """
    Server method to fetch images for a specific receipt
    This bypasses permission issues by using direct SQL query
    """
    try:
        # Verify the receipt exists and user has access
        if not frappe.db.exists("Receipt EITS", receipt_id):
            frappe.throw(_("Receipt record not found"))
        
        # Check if user has read permission for Receipt EITS
        if not frappe.has_permission("Receipt EITS", "read", receipt_id):
            frappe.throw(_("You don't have permission to access this receipt record"))
        
        # Fetch images using direct SQL query
        images = frappe.db.sql("""
            SELECT 
                name,
                image,
                remarks,
                idx
            FROM `tabImage Attachments`
            WHERE 
                parent = %s 
                AND parenttype = 'Receipt EITS'
                AND docstatus != 2
                AND image IS NOT NULL 
                AND image != '' 
                AND image != '/files/'
            ORDER BY idx ASC
        """, (receipt_id,), as_dict=True)
        
        return images
        
    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "Error in get_receipt_images")
        frappe.throw(_("Error fetching images: {0}").format(str(e)))

# Optional: Add a function to get summary data
def get_chart_data(data, filters):
    """
    Optional function to provide chart data for the report
    This creates a simple chart showing receipts by month
    """
    if not data:
        return None
    
    # Group data by month
    monthly_data = {}
    for row in data:
        if row.get('date'):
            month_key = frappe.utils.formatdate(row['date'], "MMM yyyy")
            if month_key not in monthly_data:
                monthly_data[month_key] = {'count': 0, 'amount': 0}
            monthly_data[month_key]['count'] += 1
            monthly_data[month_key]['amount'] += float(row.get('amountaed', 0))
    
    # Format for chart
    labels = list(monthly_data.keys())
    datasets = [
        {
            'name': 'Receipt Count',
            'values': [monthly_data[label]['count'] for label in labels]
        },
        {
            'name': 'Total Amount (AED)',
            'values': [monthly_data[label]['amount'] for label in labels]
        }
    ]
    
    return {
        'data': {
            'labels': labels,
            'datasets': datasets
        },
        'type': 'bar',
        'height': 300
    }

File: eits_app/eits_app/report/eits_receipt_image/eits_receipt_image.js
// JavaScript file for EITS Receipt Report
// This should be saved as the same name as your report Python file but with .js extension

frappe.query_reports["EITS Receipt Image"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 0
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 0
        },
        {
            "fieldname": "paid_by",
            "label": __("Paid By"),
            "fieldtype": "Data",
            "reqd": 0
        },
        {
            "fieldname": "paid_from",
            "label": __("Paid From"),
            "fieldtype": "Link",
            "options": "Customer",
            "reqd": 0
        },
        {
            "fieldname": "payment_method",
            "label": __("Payment Method"),
            "fieldtype": "Link",
            "options": "Method of Payment",
            "reqd": 0
        },
        {
            "fieldname": "bill_number",
            "label": __("Bill Number"),
            "fieldtype": "Data",
            "reqd": 0
        },
        {
            "fieldname": "only_with_images",
            "label": __("Only Show Records with Images"),
            "fieldtype": "Check",
            "default": 0,
            "reqd": 0
        }
    ],
    
    // Add formatter to make the Actions column clickable
    "formatter": function(value, row, column, data, default_formatter) {
        // Make Receipt ID clickable to open the document
        if (column.fieldname === "name") {
            return `<a href="/app/receipt-eits/${value}" target="_blank">${value}</a>`;
        }
        // Make View Images clickable with a proper button
        if (column.fieldname === "view_images" && value === "View Images") {
            return `<button class="btn btn-sm" onclick="view_receipt_images('${data.name}'); return false;" style="cursor: pointer;">
                        <i class="fa fa-eye"></i> View Images
                    </button>`;
        }
        // Make Submit button clickable
        if (column.fieldname === "submit_action") {
            if (value === "Submit") {
                return `<button class="btn btn-sm" onclick="submit_receipt('${data.name}'); return false;" style="cursor: pointer;">
                            <i class="fa fa-check"></i> Submit
                        </button>`;
            } else if (value === "Submitted") {
                return `<span class="badge badge-success" style="padding: 5px 10px;">
                            <i class="fa fa-check-circle"></i> Submitted
                        </span>`;
            } else if (value === "Cancelled") {
                return `<span class="badge badge-danger" style="padding: 5px 10px;">
                            <i class="fa fa-times-circle"></i> Cancelled
                        </span>`;
            }
        }
        return default_formatter(value, row, column, data);
    }
};

// Function to submit receipt - THIS WAS MISSING IN YOUR ORIGINAL CODE
function submit_receipt(receipt_id) {
    // Show confirmation dialog
    frappe.confirm(
        __('Are you sure you want to submit receipt {0}?', [receipt_id]),
        function() {
            // Show loading message
            frappe.show_alert({
                message: __('Submitting receipt...'),
                indicator: 'blue'
            });

            // Call the server method
            frappe.call({
                method: "eits_app.eits_app.report.eits_receipt_image.eits_receipt_image.submit_receipt",
                args: {
                    receipt_id: receipt_id
                },
                callback: function(r) {
                    if (r.message && r.message.success) {
                        // Show success message
                        frappe.show_alert({
                            message: r.message.message,
                            indicator: 'green'
                        });
                        
                        // Refresh the report to show updated status
                        frappe.query_report.refresh();
                        
                    } else {
                        frappe.msgprint({
                            title: __("Error"),
                            message: r.message ? r.message.message : __("Failed to submit receipt. Please try again."),
                            indicator: "red"
                        });
                    }
                },
                error: function(r) {
                    console.error("Error submitting receipt:", r);
                    let error_message = __("Error submitting receipt: ");
                    
                    if (r.exc) {
                        // Extract the actual error message from exc
                        let exc_message = r.exc;
                        if (exc_message.includes('frappe.exceptions.')) {
                            exc_message = exc_message.split(':').slice(-1)[0].trim();
                        }
                        error_message += exc_message;
                    } else if (r.message) {
                        error_message += r.message;
                    } else {
                        error_message += __("Please try again or contact administrator.");
                    }
                    
                    frappe.msgprint({
                        title: __("Error"),
                        message: error_message,
                        indicator: "red"
                    });
                }
            });
        }
    );
}

// Function to fetch and display receipt images using server method
function view_receipt_images(receipt_id) {
    // Show loading message
    frappe.show_alert({
        message: __('Loading images...'),
        indicator: 'blue'
    });

    // Call the server method we added to the Python file
    frappe.call({
        method: "eits_app.eits_app.report.eits_receipt_image.eits_receipt_image.get_receipt_images",
        args: {
            receipt_id: receipt_id
        },
        callback: function(r) {
            if (r.message && r.message.length > 0) {
                show_image_dialog(receipt_id, r.message);
            } else {
                frappe.msgprint({
                    title: __("No Images"),
                    message: __("No images found for this receipt record."),
                    indicator: "orange"
                });
            }
        },
        error: function(r) {
            console.error("Error fetching images:", r);
            frappe.msgprint({
                title: __("Error"),
                message: __("Error loading images: ") + (r.exc || "Please try again or contact administrator."),
                indicator: "red"
            });
        }
    });
}

// Function to create and show the image carousel side panel
function show_image_dialog(receipt_id, images) {
    // Filter valid images
    let validImages = images.filter(img => 
        img.image && img.image !== '/files/' && img.image.trim() !== ''
    );
    
    if (validImages.length === 0) {
        frappe.msgprint({
            title: __("No Images"),
            message: __("No valid images found for this receipt record."),
            indicator: "orange"
        });
        return;
    }
    
    // Create side panel with carousel
    create_image_carousel_panel(receipt_id, validImages);
}

// Function to create the side panel image carousel
function create_image_carousel_panel(receipt_id, images) {
    // Remove existing panel if any
    $('.image-carousel-panel').remove();
    
    let currentIndex = 0;
    
    // Create the side panel HTML
    let panelHtml = `
        <div class="image-carousel-panel" style="
            position: fixed;
            top: 0;
            right: 0;
            width: 40%;
            height: 100vh;
            background: rgba(0, 0, 0, 0.95);
            z-index: 10000;
            display: flex;
            flex-direction: column;
            box-shadow: -5px 0 15px rgba(0, 0, 0, 0.3);
        ">
            <style>
                .carousel-header {
                    background: rgba(255, 255, 255, 0.1);
                    padding: 15px 20px;
                    color: white;
                    display: flex;
                    justify-content: space-between;
                    align-items: center;
                    border-bottom: 1px solid rgba(255, 255, 255, 0.2);
                }
                .carousel-content {
                    flex: 1;
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    position: relative;
                    padding: 20px;
                }
                .carousel-image {
                    max-width: 100%;
                    max-height: 70vh;
                    object-fit: contain;
                    border-radius: 8px;
                    box-shadow: 0 4px 20px rgba(0, 0, 0, 0.5);
                }
                .carousel-nav {
                    position: absolute;
                    top: 50%;
                    transform: translateY(-50%);
                    background: rgba(255, 255, 255, 0.9);
                    border: none;
                    border-radius: 50%;
                    width: 50px;
                    height: 50px;
                    font-size: 18px;
                    cursor: pointer;
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    transition: all 0.3s ease;
                    color: #333;
                }
                .carousel-nav:hover {
                    background: white;
                    transform: translateY(-50%) scale(1.1);
                }
                .carousel-nav:disabled {
                    opacity: 0.5;
                    cursor: not-allowed;
                }
                .carousel-nav.prev {
                    left: 10px;
                }
                .carousel-nav.next {
                    right: 10px;
                }
                .carousel-footer {
                    background: rgba(255, 255, 255, 0.1);
                    padding: 15px 20px;
                    color: white;
                    border-top: 1px solid rgba(255, 255, 255, 0.2);
                }
                .image-counter {
                    background: rgba(0, 0, 0, 0.7);
                    color: white;
                    padding: 8px 15px;
                    border-radius: 20px;
                    position: absolute;
                    top: 20px;
                    left: 20px;
                    font-size: 14px;
                }
                .close-btn {
                    background: rgba(255, 255, 255, 0.2);
                    border: none;
                    color: white;
                    width: 35px;
                    height: 35px;
                    border-radius: 50%;
                    font-size: 18px;
                    cursor: pointer;
                    transition: all 0.3s ease;
                }
                .close-btn:hover {
                    background: rgba(255, 255, 255, 0.3);
                    transform: scale(1.1);
                }
                .image-remarks-panel {
                    background: rgba(255, 255, 255, 0.95);
                    color: #333;
                    padding: 10px 15px;
                    border-radius: 6px;
                    margin-top: 10px;
                    max-height: 100px;
                    overflow-y: auto;
                }
            </style>
            
            <!-- Header -->
            <div class="carousel-header">
                <div>
                    <h4 style="margin: 0; font-size: 18px;">Receipt Images: ${receipt_id}</h4>
                </div>
                <button class="close-btn" onclick="close_image_carousel()">&times;</button>
            </div>
            
            <!-- Content -->
            <div class="carousel-content">
                <div class="image-counter">
                    <span id="current-image-number">1</span> / ${images.length}
                </div>
                
                <button class="carousel-nav prev" id="prev-btn" onclick="change_image(-1)">
                    <i class="fa fa-chevron-left"></i>
                </button>
                
                <img id="carousel-image" class="carousel-image" src="${images[0].image}" alt="Receipt Image">
                
                <button class="carousel-nav next" id="next-btn" onclick="change_image(1)">
                    <i class="fa fa-chevron-right"></i>
                </button>
            </div>
            
            <!-- Footer -->
            <div class="carousel-footer">
                <div id="image-remarks">
                    ${images[0].remarks ? `
                        <div class="image-remarks-panel">
                            <strong><i class="fa fa-comment"></i> Remarks:</strong><br>
                            ${frappe.utils.escape_html(images[0].remarks)}
                        </div>
                    ` : '<div style="color: rgba(255,255,255,0.7); font-style: italic;">No remarks for this image</div>'}
                </div>
                <div style="margin-top: 10px; font-size: 12px; color: rgba(255,255,255,0.7);">
                    Use arrow keys or click buttons to navigate • ESC to close • Click outside to close
                </div>
            </div>
        </div>
    `;
    
    // Add panel to body
    $('body').append(panelHtml);
    
    // Add click handler to close when clicking outside the panel content
    $('.image-carousel-panel').on('click', function(e) {
        // Only close if clicking on the overlay (not the panel content)
        if (e.target === this) {
            close_image_carousel();
        }
    });
    
    // Prevent closing when clicking inside the panel content
    $('.carousel-panel-content').on('click', function(e) {
        e.stopPropagation();
    });
    
    // Store images data globally for navigation
    window.carouselImages = images;
    window.currentCarouselIndex = 0;
    
    // Update navigation buttons
    update_carousel_navigation();
    
    // Add keyboard navigation
    $(document).on('keydown.carousel', function(e) {
        if (e.keyCode === 37) { // Left arrow
            change_image(-1);
        } else if (e.keyCode === 39) { // Right arrow
            change_image(1);
        } else if (e.keyCode === 27) { // ESC
            close_image_carousel();
        }
    });
}

// Function to change image in carousel
function change_image(direction) {
    if (!window.carouselImages) return;
    
    let newIndex = window.currentCarouselIndex + direction;
    
    // Bounds checking
    if (newIndex < 0) newIndex = 0;
    if (newIndex >= window.carouselImages.length) newIndex = window.carouselImages.length - 1;
    
    // If index didn't change, do nothing
    if (newIndex === window.currentCarouselIndex) return;
    
    window.currentCarouselIndex = newIndex;
    let currentImage = window.carouselImages[newIndex];
    
    // Update image
    $('#carousel-image').attr('src', currentImage.image);
    
    // Update counter
    $('#current-image-number').text(newIndex + 1);
    
    // Update remarks
    let remarksHtml = currentImage.remarks ? `
        <div class="image-remarks-panel">
            <strong><i class="fa fa-comment"></i> Remarks:</strong><br>
            ${frappe.utils.escape_html(currentImage.remarks)}
        </div>
    ` : '<div style="color: rgba(255,255,255,0.7); font-style: italic;">No remarks for this image</div>';
    
    $('#image-remarks').html(remarksHtml);
    
    // Update navigation buttons
    update_carousel_navigation();
}

// Function to update navigation button states
function update_carousel_navigation() {
    if (!window.carouselImages) return;
    
    // Update prev button
    if (window.currentCarouselIndex === 0) {
        $('#prev-btn').prop('disabled', true).css('opacity', '0.5');
    } else {
        $('#prev-btn').prop('disabled', false).css('opacity', '1');
    }
    
    // Update next button
    if (window.currentCarouselIndex === window.carouselImages.length - 1) {
        $('#next-btn').prop('disabled', true).css('opacity', '0.5');
    } else {
        $('#next-btn').prop('disabled', false).css('opacity', '1');
    }
}

// Function to close the carousel panel
function close_image_carousel() {
    $('.image-carousel-panel').remove();
    $(document).off('keydown.carousel');
    window.carouselImages = null;
    window.currentCarouselIndex = 0;
}

// Additional utility function to test method paths (for debugging)
function test_submit_method(receipt_id) {
    console.log("Testing submit method for:", receipt_id);
    frappe.call({
        method: "eits_app.eits_app.report.eits_receipt_image.eits_receipt_image.submit_receipt",
        args: {
            receipt_id: receipt_id
        },
        callback: function(r) {
            console.log("Response:", r);
        },
        error: function(r) {
            console.error("Error:", r);
        }
    });
}

// Function to refresh report data
function refresh_report() {
    if (frappe.query_report) {
        frappe.query_report.refresh();
    }
}